{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "070d9180-bf81-4831-8e9a-717580c93d56",
   "metadata": {},
   "source": [
    "### Object-Oriented Programming Example (Sequence,SequenceRecord,DNASequence,AASequence)\n",
    "\n",
    "* Sequence Class \n",
    "    - Each Record should be a Sequence class\n",
    "    - all Sequence Records are in upper case and contain no number, each len of seq > 1 \n",
    "* Protein Sequence Class:\n",
    "    - Each protein sequence record contain only the standard 20 amino acids and inherits properties of Sequence\n",
    "* DNA Sequence Class: \n",
    "    - Each DNA seq record contains only 4 bases [A,T,G,C] and inherits properties of Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6dc17037-863a-431f-884a-42cd4a5392b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import total_ordering\n",
    "@total_ordering\n",
    "#making new sequence class that does not use GC content \n",
    "class Sequence():\n",
    "    def __init__(self,sequence):\n",
    "        self.sequence = sequence.upper()\n",
    "        assert len(self.sequence) > 1 , \"not a sequence\"\n",
    "    def __str__(self):\n",
    "        return f'sequence:{self.sequence}'\n",
    "    def __repr__(self):\n",
    "        x = len(self.sequence)\n",
    "        return f'sequence: {self.sequence}\\nlength : {x}'\n",
    "    def __eq__(self,other):\n",
    "        if len(self.sequence) == len(other.sequence):\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    def __lt__(self,other):\n",
    "        if len(self.sequence) < len(other.sequence):\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    def __len__(self):\n",
    "        return len(self.sequence)\n",
    "    \n",
    "class SequenceRecord():\n",
    "    def __init__(self,sequence,label):\n",
    "        assert isinstance(sequence,Sequence), \"not a Sequence object\"\n",
    "        self.sequence = sequence\n",
    "        self.label = label\n",
    "    def __str__(self):\n",
    "        output = f'{self.label}' + \"\\n\" + f'{self.sequence}'\n",
    "        return output\n",
    "    def __repr__(self):\n",
    "        output = f'>{self.label} \\n {self.sequence}'\n",
    "        return output\n",
    "class DNAsequence(Sequence):\n",
    "    def __init__(self,sequence,label):\n",
    "        Sequence.__init__(self,sequence)\n",
    "        self.sequence = sequence\n",
    "        self.label = label \n",
    "        for i in self.sequence:\n",
    "            if i not in \"ATGC\":\n",
    "                raise TypeError(\"Not a DNA Sequence\")\n",
    "    def GC(self):\n",
    "        GC = 0 \n",
    "        sequence_calc = list(self.sequence)\n",
    "        for i in sequence_calc:\n",
    "            if i == \"G\" or i == \"C\":\n",
    "                GC = GC + 1\n",
    "        self.GC = GC / len(sequence_calc)\n",
    "        return self.GC\n",
    "    def translate(self):\n",
    "        aa_dict = {'M':['ATG'], 'F':['TTT', 'TTC'], 'L':['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'C':['TGT', 'TGC'], 'Y':['TAC', 'TAT'], 'W':['TGG'], 'P':['CCT', 'CCC', 'CCA', 'CCG'], 'H':['CAT', 'CAC'],\n",
    "'Q':['CAA', 'CAG'], 'R':['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'I':['ATT', 'ATC', 'ATA'], 'T':['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "'N':['AAT', 'AAC'], 'K':['AAA', 'AAG'], 'S':['AGT', 'AGC', 'TCT', 'TCC', 'TCA', 'TCG'], 'V':['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "'A':['GCT', 'GCC', 'GCA', 'GCG'], 'D':['GAT', 'GAC'], 'E':['GAA', 'GAG'], 'G':['GGT', 'GGC', 'GGA', 'GGG'], '*':['TAA','TAG','TGA']}\n",
    "        l = []\n",
    "        q = []\n",
    "        for i in self.sequence:# loop through each char in sequence\n",
    "            l.append(i)\n",
    "            if len(l) > 2:\n",
    "                l = \"\".join(l)\n",
    "                for key,values in aa_dict.items():\n",
    "                    if l in values:\n",
    "                        q.append(key)\n",
    "                l = []\n",
    "        return \"\".join(q)\n",
    "    def __add__(self,other):\n",
    "        return (str(self.sequence)+str(other.sequence), f\"{self.label}+{other.label}\") \n",
    "    def to_fasta(self):\n",
    "        with open(f'{self.label}.fasta',\"w\") as f1:\n",
    "            f1.write(f\">{self.label}\\n\")\n",
    "            f1.write(f'{self.sequence}')\n",
    "            return f'fasta_file written to {f1.name}'\n",
    "    \n",
    "    \n",
    "class AAsequence(Sequence):\n",
    "    def __init__(self,sequence,label):\n",
    "        Sequence.__init__(self,sequence)\n",
    "        self.sequence = sequence\n",
    "        self.label = label\n",
    "        for i in self.sequence:\n",
    "            if i not in \"ACDEFGHIKLMNPQRSTVW\":\n",
    "                raise TypeError(\"Not an AA Sequence\")\n",
    "    def aliphatic_residues(self):\n",
    "        aliphatic = [\"A\",\"V\",\"M\",\"G\",\"P\",\"I\",\"L\"]\n",
    "        aliphatic_num = 0\n",
    "        for i in self.sequence:\n",
    "            if i in aliphatic:\n",
    "                aliphatic_num = aliphatic_num + 1 \n",
    "        return aliphatic_num #it would be nice to do something cooler so i might change this \n",
    "        \n",
    "    \n",
    "        \n",
    "\n",
    "            \n",
    "        \n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "af142746",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasta_file written to shh gene.fasta\n",
      "sequence:AACCATA\n",
      "sequence:AACATACACAAAAGGG\n",
      "sequence: AACCATA\n",
      "length : 7\n",
      "\n",
      "sequence: AACATACACAAAAGGG\n",
      "length : 16\n",
      "\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "seq1 = Sequence(\"AACCATA\")\n",
    "seq2 = DNAsequence(\"AACATACACAAAAGGG\",\"shh gene\")\n",
    "\n",
    "#String representation of objects \n",
    "print(seq1)\n",
    "print(seq2)\n",
    "#Repr representation \n",
    "print(f'{repr(seq1)}\\n') #repr = more verbose output than string representation \n",
    "print(f'{repr(seq2)}\\n')\n",
    "#Demostration of Polymorphism (Pillar of Object-Oriented Programming) which is a consequence of Inheritance (Pillar of Object-Oriented Programming)\n",
    "print(isinstance(seq2,DNAsequence))\n",
    "print(isinstance(seq2,Sequence))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a506e64",
   "metadata": {},
   "source": [
    "### Additional Features of Classes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "65ac4c48-4d33-4e39-9522-c2b960d2d517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence 2 (DNA)\n",
      "sequence: ATTAAAAAACA\n",
      "length : 11\n",
      "IKK\n",
      "0.09090909090909091\n",
      "----------------------------------------\n",
      "Sequence 3 (Protein)\n",
      "sequence: KTKLLKAKAKLLKVALKW\n",
      "length : 18\n",
      "9\n",
      "fasta_file written to DNA_BINDING.fasta\n"
     ]
    }
   ],
   "source": [
    "seq1 = Sequence(\"AACCATA\")\n",
    "seq2 = DNAsequence(\"ATTAAAAAACA\",\"DNA_BINDING\")\n",
    "seq3 = AAsequence(\"KTKLLKAKAKLLKVALKW\",\"Chemokine\")\n",
    "print(\"Sequence 2 (DNA)\")\n",
    "print(repr(seq2))\n",
    "print(seq2.translate()) #convert DNA --> Amino Acid Sequence\n",
    "print(f'{seq2.GC()}') #GC Content of Sequence \n",
    "print(f'{\"-\"*40}')\n",
    "print(\"Sequence 3 (Protein)\")\n",
    "print(repr(seq3))\n",
    "print(seq3.aliphatic_residues()) #returns number of Aliphatic Residues in Sequence\n",
    "\n",
    "#DNASequence object can be written to fasta file\n",
    "print(seq2.to_fasta())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
